File: fnc/examples/chapter07/qriter.m
%% 
% Let's start with a known set of eigenvalues and an orthogonal eigenvector
% basis.
D = diag([-6 -1 2 4 5]);
[V,R] = qr(randn(5));
A = V*D*V';    % note that V' = inv(V)

%%
% Now we will take the QR factorization and just reverse the factors.
[Q,R] = qr(A);
A = R*Q;

%%
% It turns out that this is a similarity transformation, so the eigenvalues
% are unchanged.
eig(A)

%%
% What's remarkable is that if we repeat the transformation many times, the
% process converges to $\mD$. 
for k = 1:15
    [Q,R] = qr(A);
    A = R*Q;
end
A

File: julia/scripts/chapter07/qriter.jl

using LinearAlgebra 
D = diagm(0=>[-6,-1,2,4,5])
V,R = qr(randn(5,5))
A = V*D*V';    # note that V' = inv(V)

Q,R = qr(A)
A = R*Q;

sort( eigvals(A) )

for k = 1:40
    Q,R = qr(A)
    A = R*Q
end
A
